### PR TITLE
URI handling in Map and Timeseries Factories

### PR DESCRIPTION
**Pull request by wmoapl**
## PR Description

Adds ability in Map and Timeseries factories to recognize and parse URI paths. It was designed with s3 paths in mind but would work for any type of URI. If path is a URI, it will use the `fsspec` package to open, read, and look for files.

#### Example code:
```
### Reads in fits file from public s3 repository using anonymous credentials
from sunpy.map import Map
Map("s3://gov-nasa-hdrl-data1/sdo/aia/20100513/0131/sdo_aia_h2_20100513T000400_0131_v1.fits", fsspec_kwargs={'anon': True})

### Reads in cdf file from public s3 repository (REQUIRES VALID AWS CREDENTIALS AND SETUP)
from sunpy.timeseries import TimeSeries
TimeSeries('s3://gov-nasa-hdrl-data1/cdaweb/ace/mag/level_2_cdaweb/mfi_k2/2017/ac_k2_mfi_20170705_v03.cdf')
```

#### Descriptions to changes in code:

- `Pyproject.toml` - adds `fsspec` as package dependency; adds `s3fs` as test dependency.
- `sunpy/io/_cdf.py` - passes kwargs to reader function.
- `Sunpy/io/_filetools.py` - adds way to open URI filetype using `fsspec.open` to `detect_filetype`
- `Sunpy/map/map_factory.py` - adds logic to separate URI file path from regular filepath and parses URIs using `parse_uri` function
- `sunpy/map/tests/test_map_factory.py` - adds unit test to read map from s3 public repository.
- `sunpy/timeseries/tests/test_timeseries_factory.py` - adds unit test to read timeseries from s3 public repository; will skip if credentials/permissions are incorrect.
- `Sunpy/timeseries/timeseries_factory.py` - adds logic to separate URI file path from regular filepath and parses URIs using parse_uri function; refactors previous code for parsing timeseries results into separate function.
- `Sunpy/util/io.py` 
  - `parse_uri` - reads in a list of fsspec OpenFile objects; follows same structure as parse_path
  - `is_uri` - function to recognize URI objects; code taken from `pandas`

Happy to iterate or discuss further on methodology.

Fixes bug in #7083

### CODE REVIEW

**Comment by nabobalis** [comment]
Hi @wmoapl, I need to review the PR but I have a quick question.

From what I understand (from the original issue) is that astropy already supports S3 URIs, we just need to add support to pass down keywords. 

Would that prevent us from needing to import and use `fsspec` directly or is that something we need to use to check if the input is a URI?

**Comment by wmoapl** [comment]
Hi @nabobalis - keywords can already be passed down to `astropy`, but the issue was that sunpy's internal parsing of paths forces it to be a local path (using `pathlib`) even if the path was a URI.

Though you do not need `fsspec` just to check if the input was a URI, I think using `fsspec` makes sense since it mimics the functionalities that are already supported for local paths, like passing directories and glob characters.

**Comment by nabobalis** [comment]
Thanks for explaining that for me. I hope to get others to review and discuss this PR during the next sunpy community meeting, so we can go forward with this.

Thanks again for the PR! Touching the IO parts of sunpy is like straightening out a mixture of cooked spaghetti.

**Comment by Cadair** [comment]
hey @wmoapl Thanks a lot for the PR. I will try and get to giving it the time it deserves soon!

File: sunpy/io/_cdf.py
Comment (wmoapl) [reply sunpy/io/_cdf.py:19]: documented all instances

File: pyproject.toml
Comment (wmoapl) [reply pyproject.toml:37]: `fsspec` is also called in `sunpy.io._file_tools.py` (used in `detect_filetypes` function) so that's why I put the dependency in core. But if there is a workaround to that, I'm happy to edit.

File: sunpy/map/tests/test_map_factory.py
Comment (Cadair) [inline sunpy/map/tests/test_map_factory.py:251]: Do you happen to know why `anon=True` needs to be specified? it seems like a sane default to me?

File: sunpy/map/tests/test_map_factory.py
Comment (Cadair) [inline sunpy/map/tests/test_map_factory.py:279]: I think rather than using these data we should use the [sunpy sample data](http://data.sunpy.org/) these are mirrored in s3 and public: I was testing with this: `m = sunpy.map.Map("s3://data.sunpy.org/sunpy/AIA20110607_065843_0193_cutout.fits", fsspec_kwargs={"anon": True})`

**Review by Cadair — COMMENTED** [review]
Hi @wmoapl

Thanks so much for this work, I think it's great. My only large concern is that there's now some duplication with this s3 support and our existing http[s] support. I think if we are going to adopt fsspec as a core dependency then we should replace all remote file handling with fsspec, and adopt it's use in more places where it makes sense (although not all of that needs to be done in this PR).

I will just ask for more input from @sunpy/sunpy-maintainers on the hard dep on fsspec, before we go down that path.

I have left a few small inline comments, but they are all minor while we think about the bigger picture.

**Comment by Cadair** [comment]
> I will just ask for more input [...] on the hard dep on fsspec, before we go down that path.

We discussed this at the community meeting yesterday, and the general consensus was that fsspec is useful enough with it's http support for us to take it on as a core dep and migrate more of our url handling over to it over time.

Please let me know if you need any more help with this PR!

File: sunpy/io/_file_tools.py
Comment (wmoapl) [reply sunpy/io/_file_tools.py]: changed

File: sunpy/map/map_factory.py
Comment (wmoapl) [reply sunpy/map/map_factory.py:191]: There is an explicit except in `is_uri` to avoid any paths that start with `http` or `https`, so line 173 and 179 should not conflict. If the path starts with `http`/`https` it will hit line 173.

File: sunpy/map/map_factory.py
Comment (wmoapl) [reply sunpy/map/map_factory.py:194]: Would you mind elaborating why you don't think it will work? It seemed to work as expected while I was testing. This was what I was running for single and multiple files:
```
single_map = sunpy.map.Map("s3://gov-nasa-hdrl-data1/sdo/aia/20100513/0131/sdo_aia_h2_20100513T000400_0131_v1.fits", fsspec_kwargs={'anon':True})
multi_map = sunpy.map.Map("s3://gov-nasa-hdrl-data1/sdo/aia/20100513/0131/sdo_aia_h2_20100513T000*_v1.fits", fsspec_kwargs={'anon':True})
```

File: sunpy/map/tests/test_map_factory.py
Comment (wmoapl) [reply sunpy/map/tests/test_map_factory.py:259]: `anon=True` is set so that the file will be read from AWS S3 without using AWS keys and only works if the file is from a public repository. I set it like this in the test scripts so that I don't have to inject AWS permissions into the test environment.

Most likely it's fine to set `anon=True` as default because most science repos are public, but it seems a bit backwards to me since it's normally not how the typical user would interact with AWS. For example if someone was trying to access files from a personal s3 repo (usually not public), `anon=True` would break.

**Comment by wmoapl** [comment]
Hi there, apologies for the slow progress on this PR. I believe I have addressed or implemented all the suggestions/comments presented. Let me know if there's anything I didn't cover.

I am having a bit of trouble passing all the CI checks. I was wondering if there is a CDF file stored somewhere on the S3 sunpy sample data repo (s3://data.sunpy.org/) or if one could be added. It would help with some additional `timeseries_factory` unit tests for the new URI handling and hopefully increase the codecov percentage to a passing level.

**Comment by nabobalis** [comment]
We can add a cdf file to the data repo which will upload it to the S3 bucket, do you have a file in mind?

**Comment by wmoapl** [comment]
> do you have a file in mind?

This [one](https://github.com/sunpy/sunpy/blob/664d1c51901edbfda9f5ac4d8f23ab3a4a4daed1/sunpy/data/test/psp_fld_l2_mag_rtn_1min_20200104_v02.cdf) would be fine.

**Comment by nabobalis** [comment]
This file has been added to the data repo. The CI says it has synced the file to S3.

File: changelog/7790.bugfix.rst
Comment (nabobalis) [inline changelog/7790.bugfix.rst:1]: I feel like this is a new feature, instead of a bugfix.

File: sunpy/map/map_factory.py
Comment (nabobalis) [inline sunpy/map/map_factory.py:205]: I am going to be a pain but could you undo all of these style changes in this and the timeseries file?

File: sunpy/timeseries/tests/test_timeseries_factory.py
Comment (nabobalis) [inline sunpy/timeseries/tests/test_timeseries_factory.py:446]: Can this match be filled in?

File: sunpy/util/io.py
Comment (nabobalis) [inline sunpy/util/io.py:130]: Maybe I missed it, are there unit tests for this function?

Is there even a unit test collection for this file?!

**Review by nabobalis — COMMENTED** [review]
Some really minor comments but otherwise for me, good to merge.

File: sunpy/util/io.py
Comment (wmoapl) [reply sunpy/util/io.py:130]: There's not an explicit test for this function but it does get called within `test_read_file_fits_uri`, `test_uri_pattern`, and `test_map_list_uri` in `test_map_factory.py` and `test_from_uri` in `test_timeseries_factory.py`.

File: sunpy/map/map_factory.py
Comment (wmoapl) [reply sunpy/map/map_factory.py]: changed

File: changelog/7790.bugfix.rst
Comment (wmoapl) [reply changelog/7790.bugfix.rst]: renamed file to `changelog/7790.feature.rst`

File: pyproject.toml
Comment (wmoapl) [reply pyproject.toml:92]: I needed to keep cdflib at 1.3.2 because it added a feature to allow anonymous reads of s3 files, which is being used in `timeseries_factory.py`

File: pyproject.toml
Comment (wmoapl) [reply pyproject.toml]: changed

File: pyproject.toml
Comment (wmoapl) [reply pyproject.toml]: tried to use the oldest versions possible and settled on these versions that were compatible:
`"boto3>=1.24.59", "s3fs>=2023.3.0",`

File: sunpy/timeseries/tests/test_timeseries_factory.py
Comment (wmoapl) [reply sunpy/timeseries/tests/test_timeseries_factory.py]: added match text

File: pyproject.toml
Comment (nabobalis) [reply pyproject.toml:92]: That's fine with me!

File: pyproject.toml
Comment (nabobalis) [reply pyproject.toml]: Thanks for working that out!

File: sunpy/util/io.py
Comment (nabobalis) [reply sunpy/util/io.py:130]: Thanks for detailing that.

File: changelog/7790.feature.rst
Comment (nabobalis) [inline changelog/7790.feature.rst:1]:
```suggestion
Added proper URI handling in `~sunpy.map.Map` and `~sunpy.timeseries.TimeSeries`.
```

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Thanks for the PR again @wmoapl. I added a whatsnew label but since it covers at least two topics, it might be easier to do that after the PR has been merged.

File: sunpy/map/map_factory.py
Comment (Cadair) [inline sunpy/map/map_factory.py:247]: I have started playing with this locally, but my main issue here is this.

Registering the `list` type is to broad (even though it should work under almost all circumstances), I would much prefer it if we flattened out the list to just be a whole bunch of `OpenFile` objects and then parsed those.

If you don't mind @wmoapl I will push a commit which does this for Map, but I haven't done it for TimeSeries yet.

**Review by Cadair — CHANGES_REQUESTED** [review]

File: sunpy/map/map_factory.py
Comment (wmoapl) [reply sunpy/map/map_factory.py:247]: Yes, that makes sense. Thanks @Cadair

Comment (Cadair) [reply sunpy/map/map_factory.py:247]: Turns out I can't push to this PR, can you merge this PR: https://github.com/APL-HTM-S3-AWARE/sunpy/pull/13

Comment (Cadair) [reply sunpy/map/map_factory.py]: I have now updated the timeseries factory as well if you can merge this: https://github.com/APL-HTM-S3-AWARE/sunpy/pull/14

**Review by Cadair — COMMENTED** [review]
Hi, I think this is now very close. I just have a few small comments and then we can merge it.

File: pyproject.toml
Comment (Cadair) [inline pyproject.toml:113]:
```suggestion
  "fsspec[s3]>=2023.3.0",
  "boto3",  # cdflib has it's own special s3 handling which requires this
```

File: pyproject.toml
Comment (Cadair) [inline pyproject.toml:114]:
```suggestion
  "boto3>=1.26.41",  # cdflib has it's own special s3 handling which requires this
```

**Review by Cadair — APPROVED** [review]

File: sunpy/map/tests/test_map_factory.py
Comment (nabobalis) [inline sunpy/map/tests/test_map_factory.py:268]: What happened here?

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Failed tests are upstream changes we can worry about in the future.

**Comment by nabobalis** [comment]
Thank you very much @wmoapl for opening this PR and sticking with us during the review process.

**Comment by Cadair** [comment]
Seconded, thanks a lot @wmoapl this is a great contribution.

File: sunpy/map/tests/test_map_factory.py
Comment (Cadair) [reply sunpy/map/tests/test_map_factory.py:268]: What do you mean? The warning is raised by the fact there's a README file in the directory which isn't a valid map.

**Comment by wmoapl** [comment]
Awesome! Thanks @nabobalis and @Cadair glad this is merged 🎉

File: sunpy/map/tests/test_map_factory.py
Comment (nabobalis) [reply sunpy/map/tests/test_map_factory.py:268]: Oh I thought it was a real file. Ignore me.

**Pull request merged**
